tests/test_async_utils.cc

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2020 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <gtest/gtest.h>
+
+#include <albatross/Indexing>
+#include <albatross/utils/AsyncUtils>
+#include <albatross/utils/RandomUtils>
+
+#include <chrono>
+#include <mutex>
+#include <numeric>
+
+namespace albatross {
+
+TEST(test_async_utils, test_async_apply_with_capture) {
+  std::vector<int> xs = {0, 1, 2, 3, 4, 5};
+
+  std::mutex mu;
+  int sum = 0.;
+
+  std::vector<int> order_processed;
+
+  auto add_to_sum = [&](const int x) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(abs(x - 2)));
+    mu.lock();
+    sum += x;
+    order_processed.push_back(x);
+    mu.unlock();
+  };
+
+  async_apply(xs, add_to_sum);
+
+  EXPECT_EQ(sum, std::accumulate(xs.begin(), xs.end(), 0));
+  // Make sure the async apply was indeed processed out of order.
+  EXPECT_NE(order_processed, xs);
```

Comment:
This could randomly fail. `std::this_thread::sleep_for` isn't guaranteed to sleep for exactly as long as you request, it could sleep longer.

Reply:
Hmmm, any other thoughts on how to test this?

Reply:
The hacky approach would be to wait until this fails and then increase the order of magnitude of sleep duration. This will also increase how long it takes to run this particular test.

I'd question if you really need to test that they executed out of order. If that's truly something that needs to be tested for it will need to be enforced in someway with synchronization primitives. I don't think the intention of this PR is to enable any kind of prioritization or order of the parallel calculations, so testing for that doesn't make a whole lot of sense.

Reply:
Yeah the out of order comparison was my attempt to test that execution was indeed happening asynchronously.  I suppose the subsequent test (where I test that execution is faster than using a single thread) accomplishes that to some degree.  I'll just remove this test for now.